### PR TITLE
Adds optional TypeAdapter support. Fixes #22

### LINKED ISSUE #22: Custom Type Adapters
Basically the same thing Gson and Moshi provide—the support for writing and reading custom not-first-hand-supported types using `Parcel`, such as `boolean`, `Date`, `Map<String, Set<String>>` and what not.

This is a proposed example for such type adapters.

```java
public final class DateAdapter implements Adapter<Date>
{
    @Override
    public void writeToParcel(@NonNull Parcel parcel, @NonNull Date date) {
        parcel.writeLong(date.getTime());
    }

    @Override
    public Date readFromParcel(@NonNull Parcel parcel) {
        return new Date(parcel.readLong());
    }
}
```

### PR DESCRIPTION
**Pull request by rharter**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,12 +23,52 @@ Simply include AutoParcel in your project and make any of your `@AutoValue` anno
 
 Now build your project and enjoy your Parcelable Foo.
 
+## TypeAdapters
+
+Out of the box AutoValue: Parcel Extension support all of the types supported by the 
+[Parcel](https://developer.android.com/reference/android/os/Parcel.html) class, but sometimes you
+may need to parcel other types, like booleans or Dates.  You can do this using a custom TypeAdapter.
```
Comment (JakeWharton) [inline README.md]: booleans are already supported by this plugin and Date is serializable. Not sure what types would be useful examples... ArrayMap? SparseLongArray?

File: adapter/src/main/java/com/ryanharter/auto/value/parcel/ParcelAdapter.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * </pre>
+ *
+ * The generated code will instantiate and use the {@code DateTypeAdapter} class to parcel and
+ * unparcel the {@code date()} property.
+ */
+@Target(METHOD)
+@Retention(SOURCE)
+@Documented
+public @interface ParcelAdapter {
+  Class<? extends TypeAdapter<?>> value();
```
Comment (JakeWharton) [inline adapter/src/main/java/com/ryanharter/auto/value/parcel/ParcelAdapter.java]: Mention that this type needs a public, no-args constructor.

File: adapter/src/main/java/com/ryanharter/auto/value/parcel/TypeAdapter.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
+package com.ryanharter.auto.value.parcel;
+
+import android.os.Parcel;
+
+/**
+ * Converts Java objects to and from Parcels.
+ *
+ * By default the AutoValue: Parcel Extension can parcel and unparcel data types known by
```
Comment (JakeWharton) [inline adapter/src/main/java/com/ryanharter/auto/value/parcel/TypeAdapter.java]: Need a `<p>` for rendered output to start a new paragraph.

File: adapter/src/main/java/com/ryanharter/auto/value/parcel/TypeAdapter.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * the {@link Parcel} class.  To support other types, like custom classes or other objects,
+ * like {@link java.util.Date} objects or booleans, you can create a custom TypeAdapter to
+ * tell the Parcel extension how to parcel the object.
+ *
+ * Here's an example TypeAdapter for a Date object:
```
Comment (JakeWharton) [inline adapter/src/main/java/com/ryanharter/auto/value/parcel/TypeAdapter.java]: Need a `<p>` for rendered output to start a new paragraph.

**Comment by rharter** [comment]
@JakeWharton updated.

**Comment by JakeWharton** [comment]
Looks good here :+1:

**Pull request merged**